lib/combinat.gi

```diff
@@ -2500,16 +2499,155 @@ InstallGlobalFunction( IteratorOfPartitionsSet, function( s )
+        iter!.next := false;
+        return [];
+      fi;
+      part := StructuralCopy(iter!.next);
+      while true do
+        j := Size(iter!.next);
+        # Compute next restricted growth string using a_i \leq Max({a_0 .. a_{i-1}})+1
+        while j > 1 do
+          # iterate in the same manner, but now we do not exceed sz-1
+          max := Minimum(Maximum(iter!.next{[1 .. j-1]}), iter!.sz-2);
+          if iter!.next[j] <= max then
+            iter!.next[j] := iter!.next[j]+1;
+            m := Maximum(iter!.next)+1;
+            break;
+          else
+            iter!.next[j] := 0;
+            j := j-1;
+          fi;
+        od;
+        m := Maximum(part)+1;
+        # Convert restricted growth string to partition of set
+        out := List([1 .. m], t -> []);
+        for i in [1 .. Size(part)] do
+          Add(out[part[i]+1], iter!.s[i]);
+        od;
+        # this is the final iteration if in the next iteration we cycle through to the start again
+        if ForAll([2 .. Size(iter!.next)], t -> iter!.next[t]=0) then
+          iter!.next := false;
+          return out;
+        fi;
+        # If the size of the partition equals sz we return, otherwise it is too small,
+        # and we continue to iterate until the next partition of the correct size is found
+        if Maximum(iter!.next)+1 = iter!.sz then
+          if m = iter!.sz then
+            return out;
+          else
+            part := StructuralCopy(iter!.next);
+          fi;
+        fi;
+      od;
+    end;
+
+    nextIteratorGivenSizeOrLess := function(iter)
+      local j, max, part, m, out, i;
+      if Length(iter!.s) = 0 or Length(iter!.s)<iter!.sz then
+        iter!.next := false;
+        return [];
+      fi;
+      part := StructuralCopy(iter!.next);
+      while true do
+        j := Size(iter!.next);
+        # Compute next restricted growth string using a_i \leq Max({a_0 .. a_{i-1}})+1
+        while j > 1 do
+          max := Minimum(Maximum(iter!.next{[1 .. j-1]}), iter!.sz-2);
+          if iter!.next[j] <= max then
+            iter!.next[j] := iter!.next[j]+1;
+            m := Maximum(iter!.next)+1;
+            break;
+          else
+            iter!.next[j] := 0;
+            j := j-1;
+          fi;
+        od;
+        m := Maximum(part)+1;
+        # Convert restricted growth string to partition of set
+        out := List([1 .. m], t -> []);
+        for i in [1 .. Size(part)] do
+          Add(out[part[i]+1], iter!.s[i]);
+        od;
+        # this is the final iteration if in the next iteration we cycle through to the start again
+        if ForAll([2 .. Size(iter!.next)], t -> iter!.next[t]=0) then
+          iter!.next := false;
+          return out;
+        fi;
+        return out;
+      od;
+    end;
+
     shallowCopy := iter -> rec( next := ShallowCopy( iter!.next ), s := iter!.s );
+    shallowCopy2 := iter -> rec( next := ShallowCopy( iter!.next ), s := iter!.s, sz := iter!.sz );
```

Comment:
I merged master into your PR and then updated the tests to match what the existing tests now do; that promptly revealed that the "new" iterators where broken with respect to `ShallowCopy`, which I fixed by adding the above `shallowCopy2` (I was too lazy to think about a better name). 

Perhaps we could unify the code again, though, by always storing `sz`, just setting it to `infinity` or `Size(s)` or `false` or whatever in the "default" case where no size is given?

Reply:
The problem with this is that the default case gives every possible partition, while with sz given it only finds partitions of that size (skipping the output of partitions which are too small, and not wasting time finding finer partitions). If you provide a default sz value it won't behave the same way as expected. You could fix this by setting the third optional argument to true (which includes the output of partitions of size less than sz) and setting the default size to Size(s) - but this is also convoluted

Reply:
I've dealt with it now in an extra commit (and after that added some further refactoring). Please have a fresh look at what I did there

Reply:
Thanks, I think it looks good, except for line 2617

    elif Length( arg ) = 1 or Length( arg ) = 2 then

which I think should be

    elif Length( arg ) > 3 then

(Sorry, I'm not sure if there was a way to reference the code directly...)

Reply:
Also, I think I misunderstood you before - we could set an arbitrary size, and it wouldn't have mattered, since it is the NextIterator function which determines how the iteration proceeds. So as long as these are set correctly it will be ok. I see this is essentially what you have done by setting it to fail. I think that is the best solution.

Reply:
You can add comments to specific lines in the diff of a PR, [see here for instructions how to do that](https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request)